component/examples/stream_example.rs

```diff
@@ -0,0 +1,144 @@
+use component::{IrrecoverableError, Manageable, Supervisor};
+use std::cmp::min;
+use std::sync::Once;
+use tokio::sync::mpsc::Sender;
+use tokio::sync::oneshot::Receiver as oneshotReceiver;
+use tokio::task::JoinHandle;
+
+static mut SHOULD_FAIL: bool = true;
+static FIXER: Once = Once::new();
+
+fn fix() {
+    FIXER.call_once(|| unsafe {
+        SHOULD_FAIL = false;
+    })
+}
+
+/// We create two structs, an empty struct as a component that will contain functions
+/// that create and fully encapsulate an instance of the actual type.
+/// An instance of the actual type needs to be instantiated inside the supervised task
+/// in order to have the correct lifetime.
+pub struct MockTcpStreamComponent {}
+
+pub struct MockTcpStream {
+    read_data: Vec<u8>,
+}
+
+impl MockTcpStream {
+    pub fn new() -> Self {
+        let read_data = Vec::new();
+        MockTcpStream { read_data }
+    }
+
+    /// This function will fail on the first call and then succeed on preceding calls to create
+    /// a situation where we have an irrecoverable error.
+    fn mock_read(&self, buf: &mut [u8]) -> Result<usize, anyhow::Error> {
+        // failure should happen once
+        unsafe {
+            if SHOULD_FAIL {
+                fix();
+                return Result::Err(anyhow!("Could not read from stream."));
+            }
+        }
+
+        let size: usize = min(self.read_data.len(), buf.len());
+        buf[..size].copy_from_slice(&self.read_data[..size]);
+        Ok(size)
+    }
+}
+
+impl MockTcpStreamComponent {
+    /// This is a function that should run continuously doing some operation, here we are
+    /// continuously listening on a mocked TCP Stream. This is ultimately the function that we
+    /// are supervising.
+    /// Inside this function we first initialize any state that will be used in this component so
+    /// that the scope is also correctly reset on a restart.
+    ///
+    /// This would be an excellent place to also add a scopeguard with a defer_panic so that if the
+    /// component panics without a user caught irrecoverable error, a descriptive error message and/
+    /// or stacktrace can be forwarded to the supervisor, and resource cleanup can occur before the
+    /// existing component goes out of scope.
+    pub async fn listen(
+        tx_irrecoverable: Sender<anyhow::Error>,
+        rx_cancellation: oneshotReceiver<()>,
+    ) {
+        // Initialize the concrete type
+        let m_tcp = MockTcpStream::new();
+
+        loop {
+            let mut buf = [0; 10];
+            let _len = match m_tcp.mock_read(&mut buf) {
+                Ok(_) => {} // process
+                Err(_) => {
+                    let e = anyhow!("missing something required");
+                    tx_irrecoverable
+                        .send(e)
+                        .await
+                        .expect("Could not send irrecoverable signal.");
+                    wait_for_cancellation(rx_cancellation).await;
```

Comment:
Why not just return here, and end it all? This relates to my questions below on the purpose of cancel why not just return an Err all the way to the root future of the task, and then get it out at the JoinHandle await?

Reply:
This is such a good question that made me stop and question many things. 

The initial ticket was explicit about using a channel to send notice of events, and it seems to be inspired by the CSP model of concurrency that is also used in Golang. 

If we could simply return an error that bubbles back up to the point of entry for these components, then we could add some simple infinite retries in main, so would we need such a library as this one? I am guessing that returning an error back up a call stack is easier said than done, or could lead to the message getting lost along the way?

Reply:
yeah, I was going to say that the easiest API for the user is to pass in an async function that returns Result, and the result is handled automatically.  I was thinking that also I thought "irrecoverable" was about expect's and unwrap's, not returning Errors which are often expected.  It looks like the model here is to explicitly send an error when we want, but I'm just afraid most people will get lazy and not implement that -- thus I think a panic catcher is important.